File: bindings/swift/Sources/Proven/SafePassword.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Safe password validation delegated to libproven FFI.
///
/// All validation and common-password checking is performed by the
/// formally verified Idris 2 core via the Zig FFI bridge.

import CProven

/// Password validation result from the FFI layer.
public struct PasswordValidationResult {
    public let score: UInt32
    public let hasUppercase: Bool
    public let hasLowercase: Bool
    public let hasDigit: Bool
    public let hasSpecial: Bool
}

public enum SafePassword {
    /// Validate a password and return its analysis (score and character class flags).
    public static func validate(_ password: String) -> Result<PasswordValidationResult, ProvenError> {
        withStringBytes(password) { ptr, len in
            let result = proven_password_validate(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(PasswordValidationResult(
                score: result.score,
                hasUppercase: result.has_uppercase,
                hasLowercase: result.has_lowercase,
                hasDigit: result.has_digit,
                hasSpecial: result.has_special
            ))
        }
    }

    /// Check if a password appears in the common passwords list.
    public static func isCommon(_ password: String) -> Bool {
        let utf8 = Array(password.utf8)
        return utf8.withUnsafeBufferPointer { buffer in
            proven_password_is_common(buffer.baseAddress, buffer.count)
        }
    }
}

File: bindings/php/src/Proven/SafePassword.php
<?php
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafePassword - FFI wrapper for proven_password_* functions.
// All computation delegated to libproven; no logic reimplemented here.

declare(strict_types=1);

namespace Proven;

/**
 * Safe password validation via libproven FFI.
 *
 * Every method calls the corresponding libproven FFI function.
 */
final class SafePassword
{
    /**
     * Validate a password and return strength analysis.
     *
     * Returns an associative array with strength, has_lowercase, has_uppercase,
     * has_digit, has_special, length on success.
     *
     * @return array{strength: int, has_lowercase: bool, has_uppercase: bool,
     *               has_digit: bool, has_special: bool, length: int}
     */
    public static function validate(string $password): array
    {
        $result = FFI::getLib()->proven_password_validate($password, strlen($password));
        return [
            'strength' => (int) $result->strength,
            'has_lowercase' => (bool) $result->has_lowercase,
            'has_uppercase' => (bool) $result->has_uppercase,
            'has_digit' => (bool) $result->has_digit,
            'has_special' => (bool) $result->has_special,
            'length' => (int) $result->length,
        ];
    }

    /**
     * Check if a password is in the common passwords list.
     */
    public static function isCommon(string $password): bool
    {
        return FFI::getLib()->proven_password_is_common($password, strlen($password));
    }
}
